### ISSUE TITLE
Extract full CSS at-rule preamble and syntax

### ISSUE THREAD

**Issue by pyoor** [w3c/reffy#1028]
Currently, the only thing related to CSS at-rules that is extracted are the descriptors associated with a particular at-rule.  The problem with this, is that ignores any syntax related to the pre-amble and body of the at-rule.  For instance, the CSS @counter-style rule, defines the following syntax:

`@counter-style <counter-style-name> { <declaration-list> }`

The `<counter-style-name>` valuespace is recorded, but not in the context of the at-rule syntax.  The `<declaration-list>` is comprised solely of descriptors in `<key>: <value>` pairs as per https://drafts.csswg.org/css-syntax-3/#typedef-declaration-list.  It'd be great if this this information is also extracted, and that the data is stored in a similar fashion to that of [mdn/data](https://github.com/mdn/data/blob/main/css/at-rules.json).  

For example:

```json
{
  "@counter-style": {
    "value": "@counter-style <counter-style-name> { <declaration-list> }",
    "descriptors": {
      "additive-symbols": {
        "for": "@counter-style",
        "value": "[ <integer> && <symbol> ]#"
      },
      "fallback": {
        "for": "@counter-style",
        "value": "<counter-style-name>"
      },
      "negative": {
        "for": "@counter-style",
        "value": "<symbol> <symbol>?"
      },
      "pad": {
        "for": "@counter-style",
        "value": "<integer> && <symbol>"
      },
      "prefix": {
        "for": "@counter-style",
        "value": "<symbol>"
      },
      "range": {
        "for": "@counter-style",
        "value": "[ [ <integer> | infinite ]{2} ]# | auto"
      },
      "speak-as": {
        "for": "@counter-style",
        "value": "auto | bullets | numbers | words | spell-out | <counter-style-name>"
      },
      "suffix": {
        "for": "@counter-style",
        "value": "<symbol>"
      },
      "symbols": {
        "for": "@counter-style",
        "value": "<symbol>+"
      },
      "system": {
        "for": "@counter-style",
        "value": "cyclic | numeric | alphabetic | symbolic | additive | [ fixed <integer>? ] | [ extends <counter-style-name> ]"
      }
    }
  }
}
```

**Comment by tidoust** [comment]
Thanks for the suggestion, @pyoor. I prepared an update to the CSS extraction code in #1033 that should take care of generating the appropriate structure (I just did not index descriptors by name).

I restricted the extraction logic to `<pre>` blocks with a `prod` class. Most CSS specs follow that convention but not all of them ([`css-counter-styles`](https://drafts.csswg.org/css-counter-styles/#the-counter-style-rule) uses a simple `<pre>` without class for instance), so some specs will need to be updated, or extraction code will need to be further amended.

On top of that, I still need to validate that the crawl yields expected results before the PR can be ready to merge, in particular that it manages to extract all at-rules (at least all at-rules that come with a formal syntax or that have descriptors) and that no descriptor gets lost in the process.

Looking at dfns extracts, the complete list of at-rules is:

- `@-webkit-keyframes` defined in `compat`
- `@keyframes` defined in `css-animations-1`
- `@import` defined in `css-cascade-3`, `css-cascade-4`, `css-cascade-5`, `CSS21`, `CSS22`
- `@layer` defined in `css-cascade-5`
- `@scope` defined in `css-cascade-6`
- `@color-profile` defined in `css-color-5`
- `@media` defined in `css-conditional-3`, `CSS21`, `CSS22`
- `@supports` defined in `css-conditional-3`
- `@when` defined in `css-conditional-5`
- `@else` defined in `css-conditional-5`
- `@container` defined in `css-contain-3`
- `@counter-style` defined in `css-counter-styles-3`
- `@viewport` defined in `css-device-adapt-1`
- `@custom-selector` defined in `css-extensions-1`
- `@font-face` defined in `css-fonts-4`, `css-fonts-5`
- `@font-feature-values` defined in `css-fonts-4`
- `@font-palette-values` defined in `css-fonts-4`
- `@nest` defined in `css-nesting-1`
- `@page` defined in `css-page-3`, `CSS21`, `CSS22`
- `@top-left-corner` defined in `css-page-3`
- `@top-left` defined in `css-page-3`
- `@top-center` defined in `css-page-3`
- `@top-right` defined in `css-page-3`
- `@top-right-corner` defined in `css-page-3`
- `@bottom-left-corner` defined in `css-page-3`
- `@bottom-left` defined in `css-page-3`
- `@bottom-center` defined in `css-page-3`
- `@bottom-right` defined in `css-page-3`
- `@bottom-right-corner` defined in `css-page-3`
- `@left-top` defined in `css-page-3`
- `@left-middle` defined in `css-page-3`
- `@left-bottom` defined in `css-page-3`
- `@right-top` defined in `css-page-3`
- `@right-middle` defined in `css-page-3`
- `@right-bottom` defined in `css-page-3`
- `@property` defined in `css-properties-values-api-1`
- `@charset` defined in `css-syntax-3`
- `media` defined in `CSS21`, `CSS22`
- `@custom-media` defined in `mediaqueries-5`
- `@scroll-timeline` defined in `scroll-animations-1`

And looking at current CSS extracts, the specs that define descriptors are:

- `compat` defines descriptors for `@media`
- `css-color-5` defines descriptors for `@color-profile`
- `css-contain-3` defines descriptors for `@container`
- `css-counter-styles` defines descriptors for `@counter-style`
- `css-device-adapt` defines descriptors for `@viewport`
- `css-fonts-5` defines descriptors for `@font-face`
- `css-fonts` defines descriptors for `@font-face`, `@font-palette-values`, `@font-feature-values`
- `css-page` defines descriptors for `@page`
- `css-properties-values-api` defines descriptors for `@property`
- `css-round-display` defines descriptors for `@media`, `@viewport`
- `mediaqueries-5` defines descriptors for `@media`
- `mediaqueries` defines descriptors for `@media`
- `scroll-animations` defines descriptors for `@scroll-timeline`

**Comment by pyoor** [comment]
@tidoust This is great!  Thanks so much for putting it together so quickly.  If there's anything I can do to help, please let me know.

**Comment by pyoor** [comment]
This is certainly something I can do.  I'll start on it today.

**Comment by pyoor** [comment]
Ok, so I reviewed the parsing.  Looks like most at-rules are parsed with the exception of `@counter-style` and `@viewport`.

| At-Rule              | Spec                        | Parsed | Notes                                                                                                                                                                                                          |
| -------------------- | --------------------------- | ------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| @counter-style       | css-counter-styles-3        | No     | Missing - @counter-style \<counter-style-name> { \<declaration-list> } |
| @viewport            | css-device-adapt-1          | No     | Missing - CSS Value definition syntax is not used here (only EBNF)                                                                                                                                             |
| @scroll-timeline     | scroll-animations-1         | No     | Not yet included in this spec.                                                                                                                                                                                 |
| @-webkit-keyframes   | compat                      | No     | Syntax is not defined in this spec.                                                                                                                                                                            |
| @font-face           | css-fonts-5                 | No     | Syntax is not defined in this spec.                                                                                                                                                                            |
| @charset             | css-syntax-3                | No     | Syntax is not defined in this spec.                                                                                                                                                                            |
| @import              | CSS21                       | No     | Syntax is not defined in this spec.                                                                                                                                                                            |
| @media               | CSS21                       | No     | Syntax is not defined in this spec.                                                                                                                                                                            |
| @page                | CSS21                       | No     | Syntax is not defined in this spec.                                                                                                                                                                            |
| @import              | CSS22                       | No     | Syntax is not defined in this spec.                                                                                                                                                                            |
| @media               | CSS22                       | No     | Syntax is not defined in this spec.                                                                                                                                                                            |
| @page                | CSS22                       | No     | Syntax is not defined in this spec.                                                                                                                                                                            |
| @keyframes           | css-animations-1            | Yes    |                                                                                                                                                                                                                |
| @import              | css-cascade-3               | Yes    |                                                                                                                                                                                                                |
| @import              | css-cascade-4               | Yes    |                                                                                                                                                                                                                |
| @layer               | css-cascade-5               | Yes    |                                                                                                                                                                                                                |
| @scope               | css-cascade-6               | Yes    |                                                                                                                                                                                                                |
| @color-profile       | css-color-5                 | Yes    |                                                                                                                                                                                                                |
| @media               | css-conditional-3           | Yes    |                                                                                                                                                                                                                |
| @supports            | css-conditional-3           | Yes    |                                                                                                                                                                                                                |
| @else                | css-conditional-5           | Yes    |                                                                                                                                                                                                                |
| @when                | css-conditional-5           | Yes    |                                                                                                                                                                                                                |
| @container           | css-contain-3               | Yes    |                                                                                                                                                                                                                |
| @custom-selector     | css-extensions-1            | Yes    |                                                                                                                                                                                                                |
| @font-face           | css-fonts-4                 | Yes    |                                                                                                                                                                                                                |
| @font-feature-values | css-fonts-4                 | Yes    |                                                                                                                                                                                                                |
| @font-palette-values | css-fonts-4                 | Yes    |                                                                                                                                                                                                                |
| @nest                | css-nesting-1               | Yes    |                                                                                                                                                                                                                |
| @bottom-center       | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @bottom-left         | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @bottom-left-corner  | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @bottom-right        | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @bottom-right-corner | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @left-bottom         | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @left-middle         | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @left-top            | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @page                | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @right-bottom        | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @right-middle        | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @right-top           | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @top-center          | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @top-left            | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @top-left-corner     | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @top-right           | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @top-right-corner    | css-page-3                  | Yes    |                                                                                                                                                                                                                |
| @property            | css-properties-values-api-1 | Yes    |                                                                                                                                                                                                                |
| @custom-media        | mediaqueries-5              | Yes    |                                                                                                                                                                                                                |

**Comment by pyoor** [comment]
I'll see if I can fix the parsing to match the `@counter-style` rule.

**Comment by pyoor** [comment]
Rather than modifying the parsing, I've opened https://github.com/w3c/csswg-drafts/issues/7565 to correct the spec.

**Comment by tidoust** [comment]
Many thanks for the analysis and the PR @pyoor!

I was surprised to see that `@scroll-timeline` no longer shows up but the scroll linked animations has indeed [recently been re-written](https://github.com/w3c/csswg-drafts/commit/fc6766d95aeb0768ec8a09253d73114090369614) with a new declarative model that drops the at-rule.

**Comment by tidoust** [comment]
Released in v9.0.0 of Reffy, and the result is now available in Webref. Thanks again for the help, @pyoor!
